### PR TITLE
[ci] Check for timestamps in Jenkinsfile changes

### LINKED ISSUE #12388: [ci] Assert that Jenkinsfile changes also change the timestamp
Concurrent Jenkinsfile changes that happen in a sequence like:

1. PR 1 makes changes
2. PR 2 makes changes
3. PR 1 merges
4. PR 2 merges

Can create a situation where both PRs pass CI but break `main` when merged with failures like https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/main/4024/pipeline.

There is a timestamp in the `Jenkinsfile` that makes it so these PRs will have a merge conflict (in the example above PR 2 must be rebased and run through CI again before it can merge), we should check that any changes to the Jenkinsfile also change this timestamp (so if say PR 2 did not change the timestamp it wouldn't pass PR CI)

cc @Mousius @areusch @gigiblender

### PR DESCRIPTION
**Pull request by driazati**
This closes #12388

cc @Mousius @areusch @gigiblender

### CODE REVIEW

**Review by Mousius — CHANGES_REQUESTED** [review]
Hi @driazati, do we have any tests around this script? This logic seems to be complex enough to warrant some unit tests?

File: ci/jenkins/generate.py [hunk 1]
```diff
@@ -72,12 +75,49 @@
...
+def check_timestamp() -> None:
+    """
+    Assert that the git diff against main contains a timestamp
+    """
+    files = changed_files()
+
+    # Check the Jenkinsfile timestamp if the templates were edited
+    if any(fnmatch.fnmatch(f, "ci/jenkins/*.j2") for f in files):
+        proc = subprocess.run(
+            ["git", "diff", "origin/main", "HEAD"],
```
Comment (areusch) [inline ci/jenkins/generate.py:105]: shall we limit to Jenkinsfile?

**Review by areusch — COMMENTED** [review]
thanks @driazati , agree with @Mousius about tests, and added a couple small comments.

File: tests/python/ci/test_ci.py [hunk 2]
```diff
@@ -1100,5 +1101,72 @@ def test_should_rebuild_docker(tmpdir_factory, changed_files, name, check, expec
...
+    git.run("add", ".")
+    git.run("commit", "-mtest")
+
+    # Run generate.py in --check mode
+    proc = subprocess.run(
+        [
+            str(script),
+            "--check",
+            "--base-ref",
+            "HEAD~1",
```
Comment (areusch) [inline tests/python/ci/test_ci.py:1159]: i think HEAD~1 won't catch everything right, since you could have multiple commits in a Jenkinsfile PR?

Comment (areusch) [reply tests/python/ci/test_ci.py:1159]: sorry, just realized i'm looking at the test, ignore.

**Review by areusch — APPROVED** [review]

**Pull request closed without merging**